crates/store/re_chunk/src/iter.rs

```diff
@@ -1,29 +1,179 @@
 use std::sync::Arc;
 
-use arrow2::array::Array as ArrowArray;
-use itertools::Itertools as _;
+use arrow2::{
+    array::{Array as ArrowArray, PrimitiveArray},
+    Either,
+};
+use itertools::izip;
 
 use re_log_types::{TimeInt, Timeline};
-use re_types_core::ComponentName;
+use re_types_core::{Component, ComponentName};
 
 use crate::{Chunk, ChunkTimeline, RowId};
 
 // ---
 
+impl Chunk {
+    /// Returns an iterator over the indices (`(TimeInt, RowId)`) of a [`Chunk`], for a given timeline.
+    ///
+    /// If the chunk is static, `timeline` will be ignored.
+    ///
+    /// See also:
+    /// * [`Self::iter_component_indices`].
+    /// * [`Self::iter_indices_owned`].
+    #[inline]
+    pub fn iter_indices(&self, timeline: &Timeline) -> impl Iterator<Item = (TimeInt, RowId)> + '_ {
+        if self.is_static() {
+            Either::Right(Either::Left(izip!(
+                std::iter::repeat(TimeInt::STATIC),
+                self.row_ids()
+            )))
+        } else {
+            let Some(time_chunk) = self.timelines.get(timeline) else {
+                return Either::Left(std::iter::empty());
+            };
+
+            Either::Right(Either::Right(izip!(time_chunk.times(), self.row_ids())))
+        }
+    }
+
+    /// Returns an iterator over the indices (`(TimeInt, RowId)`) of a [`Chunk`], for a given
+    /// timeline and component.
+    ///
+    /// If the chunk is static, `timeline` will be ignored.
+    ///
+    /// This is different than [`Self::iter_indices`] in that it will only yield indices for rows
+    /// at which there is data for the specified `component_name`.
+    ///
+    /// See also [`Self::iter_indices`].
+    pub fn iter_component_indices(
+        &self,
+        timeline: &Timeline,
+        component_name: &ComponentName,
+    ) -> impl Iterator<Item = (TimeInt, RowId)> + '_ {
+        let Some(list_array) = self.components.get(component_name) else {
+            return Either::Left(std::iter::empty());
+        };
+
+        if self.is_static() {
+            let indices = izip!(std::iter::repeat(TimeInt::STATIC), self.row_ids());
+
+            if let Some(validity) = list_array.validity() {
+                Either::Right(Either::Left(Either::Left(
```

Comment:
... not sure why I whined about double nested either earlier Oo. Maybe you want to introduce new `IterVariants3` and `IterVariants4` iterators which internally match things?

Reply:
Tbh I'd rather keep the nicely composable `Either` than introduce more ad-hoc things.

The caller never has to deal with them anyhow :shrug: